### PR TITLE
3SI Transaction test cases scenario 02 #38

### PR DESCRIPTION
**Pull request by YazhiFeng**
code functionality tested correct. 
But leave two question as listed on https://github.com/futurewei-cloud/chogori-platform/issues/38. 
The code is submitted in case you need to check the code.

### CODE REVIEW

**Comment by YazhiFeng** [comment]
All scenario-02 cases are implemented and the "failed_assignment" bug is fixed. Please check the PR.

File: test/k23si/3SITxnTest.cpp [hunk 1]
```diff
@@ -1279,6 +1285,484 @@ seastar::future<> testScenario01() {
...
+                .then([this, &k6, &eqMTR] {
+                    dto::K23SI_MTR bfMTR{
+                        .txnid = txnids++, 
+                        .timestamp = {(eqMTR.timestamp.tEndTSECount() - 500000000), 123, 1000}, // 500ms earlier
+                        .priority = dto::TxnPriority::Medium};
+                    return doRead<DataRec>(k6, bfMTR, collname, ErrorCaseOpt::NoInjection)
+                    .then([](auto&& response) {
+                        auto& [status, resp] = response;
+                        K2INFO("SC02.case08-09-10(before)::OP_READ. " << "status: " << status.code << " with MESG: " << status.message);
+                        K2EXPECT(status, Statuses::S404_Not_Found);
```
Comment (jfunston) [inline test/k23si/3SITxnTest.cpp:1668]: Here and elsewhere you can use the statuses from K23SIStatus in dto/K23SI.h which is a bit more clear in intent

File: test/k23si/3SITxnTest.cpp [hunk 2]
```diff
@@ -1279,6 +1285,484 @@ seastar::future<> testScenario01() {
...
+                        K2INFO("Value of Key6: " << resp.value.val);
+                        K2EXPECT(status, Statuses::S200_OK);                        
+                        K2EXPECT(resp.value.val, v1);
+                    });
+                }); // end sc-02 case-10
+            }
+        );
+    }) // end sc-02 case-08-09-10
+    .then([] {
+    // SC02 case11 & 12 & 13 & 14: Async END test (end-but-not-finalized key )
```
Comment (jfunston) [inline test/k23si/3SITxnTest.cpp:1699]: Can you explain a bit more in the comment what the purpose of this test case is?

File: test/k23si/3SITxnTest.cpp [hunk 3]
```diff
@@ -1279,6 +1285,484 @@ seastar::future<> testScenario01() {
...
+                    return seastar::when_all(doFinalize(k11, k11, commitMTR, collname, false, ErrorCaseOpt::NoInjection), doFinalize(k12, k12, abortMTR, collname, true, ErrorCaseOpt::NoInjection))
+                    .then([&](auto&& response) mutable {
+                        auto& [resp1, resp2] = response;
+                        // move resp out of the incoming futures sice get0() returns an rvalue
+                        auto [status1, val1] = resp1.get0();
+                        auto [status2, val2] = resp2.get0();
+                        K2INFO("SC01.case11::OP_Finalize(abort)_a_END(commit)_key. " << "status: " << status1.code << " with MESG: " << status1.message);
+                        K2INFO("SC01.case11::OP_Finalize(commit)_a_END(abort)_key. " << "status: " << status2.code << " with MESG: " << status2.message);
+                        K2EXPECT(status1, Statuses::S405_Method_Not_Allowed);
+                        K2EXPECT(status2, Statuses::S405_Method_Not_Allowed);
```
Comment (jfunston) [inline test/k23si/3SITxnTest.cpp:1744]: What do we get 405 here? I would think it should succeed. Am I misunderstanding it?

**Review by jfunston — CHANGES_REQUESTED** [review]
You will also need to rebase on master since we merged the SKV interface which changes the READ and WRITE dtos. I made the changes to your previous test cases, you can refer to them to see how to update the code.

**Comment by jfunston** [comment]
Looks good, once you get it rebased on latest master with the new SKV interface it will be good to merge.

**Review by jfunston — APPROVED** [review]

**Pull request merged**